### ISSUE TITLE
lsyncd-2.2.1 crashes when attempting to sync directory-tree deletions

### ISSUE THREAD

**Issue by ferdnyc** [axkibe/lsyncd#422]
Fedora has a build of lsyncd-2.2.1 in testing (to replace the lsyncd-2.1.5 we've been running with for almost a year now), and we appear to be seeing a regression with delete-syncing. You can see full details at [Fedora bug#1415295](https://bugzilla.redhat.com/show_bug.cgi?id=1415295) if you like, but I'm including a summary of the bug and a simple testcase to reproduce it below.

When lsyncd-2.2.1 attempts to sync the deletion of a (populated) directory tree, it crashes with the following error:

````
file has vanished: "/full/path/to/lsyncd-sourcedir/some/directory"
ABORTING due to invalid path from sender: some/directory/file.inner
rsync error: protocol incompatibility (code 2) at generator.c(1271) [generator=3.1.2]
05:05:09 Error: Failure (list): 2
05:05:09 Error: Critical exitcode.
````

This occurs reproducibly (simple testcase follows), and does not occur with the previous lsyncd-2.1.5 build, both using the very same `rsync  version 3.1.2  protocol version 31` underneath the hood.

### To Reproduce

1. `mkdir /var/tmp/lsynctest` on both source and destination
2. Start lsyncd configured as follows:
````
settings {
	logfile = "/var/tmp/lsynctest.log",
	statusFile = "/var/tmp/lsynctest.pid",
	maxProcesses = 1,
}

sync{
	default.rsyncssh,
	source="/var/tmp/lsynctest/",
	host="YOUR_REMOTEHOST",
	targetdir="/var/tmp/lsynctest/",
	delete = true,
	rsync = {
	  archive = true,
	  compress = false,
	  whole_file = false,
	  hard_links = true,
	  update = true,
	}
}
````
3. Once lsyncd has finished startup, `cp -R /etc/systemd /var/tmp/lsynctest/`
    (Or any other directory tree, as long as it has depth. We need a populated subdirectory to delete.)
4. Wait for lsyncd to sync the tree.
5. `rm -r /var/tmp/lsynctest/systemd/system/multi-user.target.wants`
6. When lsyncd attempts to sync the directory deletion...

````
05:05:09 Normal: Rsyncing list
/systemd/system/multi-user.target.wants/abrtd.service
/systemd/system/multi-user.target.wants/rpcbind.service
/systemd/system/multi-user.target.wants/avahi-daemon.service
/systemd/system/multi-user.target.wants/auditd.service
/systemd/system/multi-user.target.wants/rsyslog.service
/systemd/system/multi-user.target.wants/akmods-shutdown.service
/systemd/system/multi-user.target.wants/atd.service
/systemd/system/multi-user.target.wants/hddtemp.service
/systemd/system/multi-user.target.wants/ModemManager.service
/systemd/system/multi-user.target.wants/crond.service
/systemd/system/multi-user.target.wants/sshd.service
/systemd/system/multi-user.target.wants/remote-fs.target
/systemd/system/multi-user.target.wants/restorecond.service
/systemd/system/multi-user.target.wants/gpm.service
/systemd/system/multi-user.target.wants/netcf-transaction.service
/systemd/system/multi-user.target.wants/sysstat.service
/systemd/system/multi-user.target.wants/sensord.service
/systemd/system/multi-user.target.wants/abrt-ccpp.service
/systemd/system/multi-user.target.wants/abrt-vmcore.service
/systemd/system/multi-user.target.wants/lm_sensors.service
/systemd/system/multi-user.target.wants/mcelog.service
/systemd/system/multi-user.target.wants/sendmail.service
/systemd/system/multi-user.target.wants/cups.path
/systemd/system/multi-user.target.wants/smartd.service
/systemd/system/multi-user.target.wants/abrt-xorg.service
/systemd/system/multi-user.target.wants/chronyd.service
/systemd/system/multi-user.target.wants/mdmonitor.service
/systemd/system/multi-user.target.wants/NetworkManager-dispatcher.service
/systemd/system/multi-user.target.wants/sm-client.service
/systemd/system/multi-user.target.wants/NetworkManager.service
/systemd/system/multi-user.target.wants/libvirtd.service
/systemd/system/multi-user.target.wants/irqbalance.service
/systemd/system/multi-user.target.wants/abrt-oops.service
/systemd/system/multi-user.target.wants/akmods.service
/systemd/system/multi-user.target.wants/smb.service
/systemd/system/multi-user.target.wants
file has vanished: "/var/tmp/lsynctest/systemd/system/multi-user.target.wants"
ABORTING due to invalid path from sender: systemd/system/multi-user.target.wants/ModemManager.service
rsync error: protocol incompatibility (code 2) at generator.c(1271) [generator=3.1.2]
05:05:09 Error: Failure (list): 2
05:05:09 Error: Critical exitcode.
````
The way I read that error, it _appears_ that the parent directory on the remote end is being deleted while rsync is still accessing the files inside it. Is that possible? Is this an `lsyncd-2.2.1` bug, or an issue with Fedora's build? It definitely doesn't occur with Fedora's previous `lsyncd-2.1.5`.

**Comment by axkibe** [comment]
Works for me.

I get
```
file has vanished: "/home/axel/src/systemd/system/multi-user.target.wants"
rsync warning: some files vanished before they could be transferred (code 24) at main.c(1183) [sender=3.1.1]
14:02:32 Normal: Finished (list): 24
```

with exitcode 24 from rsync is considered normal by Lsyncd.

Googling for your error message from rsync: "rsync error: protocol incompatibility (code 2) at generator.c(1271)"

This is normally caused by .bashrc on the target machine that interfers with the rsync protocol.

**Comment by axkibe** [comment]
I just got the same error as use by release testing.

This needs fixing, not yet sure if it's a rsyncd issue or if I can workaround this from Lsyncd.

**Comment by axkibe** [comment]
I just released 2.2.2

which worksaround this issue by reverting to use old style exclusion filters instead of file lists like 2.2.0 and 2.2.1 did.

Once rsync fixes this, Lsyncd can go again with lists.

**Comment by axkibe** [comment]
Closing this as duplicate of #423
